datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
+        turns out to stay that way, we should likely fuse this with
+        `self.files_all`/`self.files_latest` and not keep all of it in memory.
+        """
+
+        if self._dataset is None:
+            # This delivers a full record of all known versions of this dataset.
+            # Hence, the file lists in the version entries may contain
+            # duplicates (unchanged files across versions).
+            versions = self.api.get_dataset_versions(self.doi)
+            versions.raise_for_status()
+
+            self._dataset = versions.json()['data']
+            # Expected structure in self._dataset is a list of (version-)
+            # dictionaries, which should have a field 'files'. This again is a
+            # list of dicts like this:
+            #  {'description': '',
+            #   'label': 'third_file.md',
+            #   'restricted': False,
+            #   'directoryLabel': 'subdir2',
+            #   'version': 1,
+            #   'datasetVersionId': 72,
+            #   'dataFile': {'id': 682,
+            #   'persistentId': '',
+            #   'pidURL': '',
+            #   'filename': 'third_file.md',
+            #   'contentType': 'text/plain',
+            #   'filesize': 9,
+            #   'description': '',
+            #   'storageIdentifier': 'local://1821bc70e68-c3c9dedcfce6',
+            #   'rootDataFileId': -1,
+            #   'md5': 'd8d77109f4a24efc3bd53d7cabb7ee35',
+            #   'checksum': {'type': 'MD5',
+            #                'value': 'd8d77109f4a24efc3bd53d7cabb7ee35'},
+            #   'creationDate': '2022-07-20'}
+
+            # sort by version, so we can rely on first entry to refer to latest
+            # Note, that ('versionNumber', 'versionMinorNumber', 'versionState')
+            # would look like this:
+            # (None, None, 'DRAFT'), (2, 0, 'RELEASED'), (1, 0, 'RELEASED')
+            # and we need a possible DRAFT to have the biggest key WRT sorting.
+            self._dataset.sort(key=lambda v:
+                                   (v.get('versionNumber') or sys.maxsize,
+                                    v.get('versionMinorNumber') or sys.maxsize),
+                               reverse=True)
+
+        return self._dataset
+
+    @property
+    def files_all(self):
+        """Cache of files available from dataverse.
+
+        For quick lookup and deduplication, this is a dict {id: remote_path}
+        where remote_path is a `Path` build from the file's 'directoryLabel'
+        (if any) and 'filename'.
+        Currently, this is all we need, hence don't bother carrying and updating
+        all other metadata.
+        """
+
+        if self._files_all is None:
+            self._files_all = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f_l in [version['files'] for version in self.dataset]
+                for f in f_l}
```

Comment:
Note to self: The dict covers all files in all version of the dataset that ever existed.